cmd/gcp-controller-manager/kubelet_readonly_csr_approver.go

```diff
@@ -0,0 +1,327 @@
+	kubeletReadonlyCSRRequestMetrics     = "kubelet_readonly_csr_request"
+	podNameKey                           = "authentication.kubernetes.io/pod-name"
+	kubeletAPILimitedReaderAnnotationKey = "autopilot.gke.io/kubelet-api-limited-reader"
+)
+
+var (
+	kubeletReadonlyUsageAllowList = []capi.KeyUsage{
+		capi.UsageKeyEncipherment,
+		capi.UsageDigitalSignature,
+		capi.UsageClientAuth,
+	}
+)
+
+type kubeletReadonlyCSRRequestContent struct {
+	context           context.Context
+	controllerContext *controllerContext
+	csr               *capi.CertificateSigningRequest
+	x509cr            *x509.CertificateRequest
+}
+
+type kubeletReadonlyCSRRequestResult struct {
+	result  bool
+	err     error
+	message string
+}
+
+type readonlyResourceCSRValidateFunc func(content kubeletReadonlyCSRRequestContent) kubeletReadonlyCSRRequestResult
+
+// readonlyApprover handles approval/denial of CSRs access to read only resource
+// There is a list of approvers, requst must get all approval
+type readonlyResourceApprover struct {
+	ctx        *controllerContext
+	validators []readonlyResourceCsrValidator
+}
+
+func newKubeletReadonlyCSRApprover(ctx *controllerContext) readonlyResourceApprover {
+	return readonlyResourceApprover{
+		ctx:        ctx,
+		validators: newReadonlyResourceCsrValidator(),
+	}
+}
+
+func (approver *readonlyResourceApprover) handle(ctx context.Context, csr *capi.CertificateSigningRequest) error {
+	if csr == nil || csr.Spec.SignerName != kubeletReadonlyCSRSignerName {
+		return nil
+	}
+	klog.Infof("approver got CSR %q", csr.Name)
+	recordMetric := csrmetrics.ApprovalStartRecorder(kubeletReadonlyCSRRequestMetrics)
+	if len(csr.Status.Certificate) != 0 {
+		return nil
+	}
+	if approved, denied := certificates.GetCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+
+	x509cr, err := certutil.ParseCSR(csr.Spec.Request)
+	if err != nil {
+		recordMetric(csrmetrics.ApprovalStatusParseError)
+		klog.Errorf("unable to parse csr %q: %v", csr.Name, err)
+		return fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
+	}
+
+	for _, validator := range approver.validators {
+		recordValidatorMetric := csrmetrics.ApprovalStartRecorder(validator.authFlowLabel)
+		content := kubeletReadonlyCSRRequestContent{
+			context:           ctx,
+			controllerContext: approver.ctx,
+			csr:               csr,
+			x509cr:            x509cr,
+		}
+		result := validator.validate(content)
+
+		switch {
+		case result.err != nil:
+			klog.Errorf("validating CSR %q: %v", csr.Name, err)
+			return fmt.Errorf("validating CSR %q: %v", csr.Name, err)
+		case result.result == false:
+			klog.Infof("validator %q: denied CSR %q", validator.name, csr.Name)
+			recordValidatorMetric(csrmetrics.ApprovalStatusDeny)
+			return approver.updateCSR(csr, false, result.message)
+		case result.result == true:
+			klog.Infof("validator %q: approved CSR %q", validator.name, csr.Name)
+			recordValidatorMetric(csrmetrics.ApprovalStatusApprove)
```

Comment:
you should only approve if all of the validators return true.

Reply:
Good point